### PR TITLE
Add support for opening the database with TTL support

### PR DESCRIPTION
**Pull request by silviucpp**

### CODE REVIEW

File: src/erocksdb.erl
Comment (mocchira) [inline src/erocksdb.erl]: Upon our coding convention,
It would be more desirable to name _TTL rather than _Ttl. (We'd expect that it would be more readable for many devs that TTL means "time to live" however Ttl might not work as well)

File: src/erocksdb.erl
Comment (mocchira) [inline src/erocksdb.erl]: ditto

File: src/erocksdb.erl
Comment (mocchira) [inline src/erocksdb.erl]: ditto

File: c_src/workitems.cc
Comment (mocchira) [inline c_src/workitems.cc:172]: Please tell us the reason why you've decided to adopt this behavior (if ttl is set to 0 or less than 0 then fallback to the normal rocksdb::DB::Open).

According to the official document at https://github.com/facebook/rocksdb/wiki/Time-to-Live, rocksdb behave as TTL = infinity in such cases.

Without any comment on the code, erocksdb users might expect the code to behave what the official document says.

That  being said, if there are some rational reason to fallback then we'd have to document/comment about it on the comment of function open/4 at src/erocksdb.erl.

File: src/erocksdb.erl
Comment (mocchira) [inline src/erocksdb.erl]: number() is not the appropriate type because which covers both integer() and float(). integer() seems to be suitable according to the corresponding rocksdb function (DBWithTTL::Open) spec.

**Review by mocchira — CHANGES_REQUESTED** [review]
Basic logic looks good. there are a few suggestions and confirmation about the rational reason behind the design decision you made.

And also I'd like to ask you to add a test suite for open/4 similar to the one for open/3 https://github.com/leo-project/erocksdb/blob/develop/src/erocksdb.erl#L612-L624 in order to confirm the basic function could work as expected.

File: c_src/workitems.cc
Comment (silviucpp) [reply c_src/workitems.cc:172]: hmm .. infinity doesn't mean should behave like the default open function ? This was my first thought when I readed the documentation. That in case you open using rocksdb::DBWithTTL::Open having TTL <= 0 it's same stuff like using rocksdb::DB::Open where records are never removed based on any TTL..

File: src/erocksdb.erl
Comment (silviucpp) [reply src/erocksdb.erl]: Agree will change this

File: src/erocksdb.erl
Comment (silviucpp) [reply src/erocksdb.erl]: Clear will change all this Ttl into TTL

File: c_src/workitems.cc
Comment (mocchira) [reply c_src/workitems.cc:172]: Yes partly I think so.
Although I've not confirmed yet, There is a little bit possibility that a internal key format might differ from the original one when TTL <= 0 according to the spec

- (int32_t)Timestamp(creation) is suffixed to values in Put internally
- Not specifying/passing or non-positive TTL behaves like TTL = infinity

Depending on how rocksdb implement TTL, each key might be suffixed by Timestamp(creation) regardless of the value of TTL. 

if this is the case then we can't do this with the current PR

- open a file with TTL <= 0
- put some records
- close the file
- open the file with TTL > 0

because the internal key format of records being used during the first open is not suffixed by Timestamp(creation).

and also

> That in case you open using rocksdb::DBWithTTL::Open having TTL <= 0 it's same stuff like using rocksdb::DB::Open where records are never removed based on any TTL.

if this is true then we can rewrite the code like below (no branch by ttl <= 0) to make the code simpler.

```cpp
rocksdb::DBWithTTL* db_with_ttl(0);
status = rocksdb::DBWithTTL::Open(*options, db_name, &db_with_ttl, ttl);
db = db_with_ttl;
```

**Review by mocchira — CHANGES_REQUESTED** [review]
LGTM other than how we should open a TTL file. Please check the latest comment out.

File: c_src/workitems.cc
Comment (silviucpp) [reply c_src/workitems.cc:172]: Hello @mocchira, you are right. I mean I agree we shouldn't depend on the rocksdb internals, and better we should find another way to decide when we open with TTL and when we don't . 

What do you think if in API I will add another function:

open_with_ttl/3 and internally to native code I will send a tuple like {use_ttl, TTL}.

For regular open will send {false, undefined} and for open_with_ttl/3 {true, integer()}.

Or any other idea ?

Silviu

Comment (mocchira) [reply c_src/workitems.cc:172]: > open_with_ttl/3 and internally to native code I will send a tuple like {use_ttl, TTL}.

Sounds good :)
Please go ahead with this direction.

**Comment by silviucpp** [comment]
@mocchira  I added the dedicated API for open_with_ttl and did the changes we talked about. please review

File: c_src/workitems.cc
Comment (mocchira) [inline c_src/workitems.cc]: Use 4 spaces instead of 1 tab for the indent.

**Review by mocchira — CHANGES_REQUESTED** [review]
LGTM. Only one thing about the indent should be fixed. Once the PR get fixed, I'll merge it. Thanks.

File: c_src/workitems.cc
Comment (silviucpp) [reply c_src/workitems.cc]: Are you sure it's a tab there ? Because for me shows 4 spaces

Comment (mocchira) [reply c_src/workitems.cc]: Sorry I pointed out the wrong file.
please check "bool use_ttl;" in c_src/workitems.h
here a tab is used as an indent.

File: c_src/workitems.h
Comment (mocchira) [inline c_src/workitems.h]: Here a tab is used as an indent.

**Review by mocchira — CHANGES_REQUESTED** [review]

**Comment by silviucpp** [comment]
Done

**Review by mocchira — APPROVED** [review]

**Comment by mocchira** [comment]
@silviucpp LGTM. Thanks for your contribution. I will merge once the travis-ci get the job done.

**Pull request merged**
